### ISSUE TITLE
[CUDA] GLM-5.2/DeepseekV32 mxfp4 forward numerically wrong on Blackwell (sm_120/121/110), correct on Metal — individual ops fine, composed forward diverges ~25%

### ISSUE THREAD

**Issue by Stefanfindeisen** [ml-explore/mlx#3879]
## Summary

The **GLM-5.2 (`mlx-community/GLM-5.2-mxfp4`, `model_type: glm_moe_dsa`, DeepseekV32-style MLA + 256-expert MoE) forward is numerically wrong on Blackwell CUDA but correct on Metal.**

Running the **same** 4-layer slice of the model (3 dense + 1 MoE/DSA layer) with the **real mxfp4 weights loaded from disk**, on the **same inputs**, comparing GPU vs CPU:

| device | rel-diff vs CPU |
|---|---|
| Apple M5 Max (Metal) | **0.033** (correct) |
| RTX 5090 (`sm_120`, CUDA 12.8) | **0.25** (broken) |

(Same pattern on GB10 `sm_121` and Jetson Thor `sm_110` in the live cluster.) Over 78 layers this compounds into degenerate logits — distributed inference across a mixed Metal+Blackwell ring produces pure garbage (`"!"` repeated), while a standard int4 model (Qwen3) on the identical ring produces coherent text.

## The puzzle: individual ops are fine, the composed forward is not

On the same `sm_120` device, GPU-vs-CPU for the building blocks is all within quantization noise:

- `mx.quantized_matmul(..., mode='mxfp4')` across many shapes (2048×6144, 6144×6144, 1536×6144, 512×6144, 6144×512, 154880×6144, …): **0.010–0.031**
- `mx.fast.rope`: 0.0014 · `mx.fast.scaled_dot_product_attention` (+causal): 0.003–0.004 · `mx.fast.rms_norm`: 0.0000

Yet the **composed DeepseekV32 dense-layer forward** diverges **0.22–0.25** on Blackwell (and only 0.033 on Metal). So it is not any single op in isolation — it is Blackwell-specific behavior of the composed graph (possibly error amplification through the deep MLA attention chain, a CUDA-graph precision issue, or an MLA op I haven't isolated that only misbehaves in-model).

## Environment

- mlx `0.32.0.dev20260717+7a1d4f5`, source-built with `-DMLX_BUILD_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=120/121/110`.
- RTX 5090 (sm_120, CUDA 12.8, WSL2); GB10 Spark (sm_121, CUDA 13.0); Jetson Thor (sm_110, JetPack 7.1).
- Metal reference: M5 Max / M4 Max.

I have standalone repro scripts (load 4 real GLM layers, forward GPU vs CPU) and can run any instrumented build or narrower repro on the Blackwell hardware. Happy to help bisect which op in the MLA/DeepseekV32 path diverges.

Related: #3876 (distributed barrier deadlock on the same cluster/arch).

**Comment by Stefanfindeisen** [comment]
## Narrowed: it's error amplification in **chained** mxfp4 ops, not any single op

Ran identical GPU-vs-CPU scripts on Blackwell (`sm_120`) and Metal (M5), building up from single ops:

| test | Metal (M5) | Blackwell (sm_120) |
|---|---|---|
| single `quantized_matmul` mxfp4 (any out_dim 2048–12288, any weight scale, gs=32) | 0.016 | 0.016 |
| `silu` / `rope` / `sdpa` / `rms_norm` | ~0 | <0.004 |
| **SwiGLU MLP** (`down(silu(gate(x))*up(x))`, all mxfp4) | **0.035** | **0.258** |
| real GLM dense-layer sub-modules | — | attn 0.11, mlp 0.17 |
| real GLM forward (4 layers) | 0.033 | 0.25 |

So **a single mxfp4 matmul is fine on Blackwell (0.016, same as Metal)** — but as soon as you **chain** mxfp4 matmuls through an activation (matmul → silu·mul → matmul), Blackwell blows up to 0.26 while Metal stays at 0.035. Over 78 layers this compounds into fully degenerate logits.

### Minimal repro (Blackwell vs Metal, same script)

```python
import mlx.core as mx, mlx.nn as nn
mx.random.seed(0)
h, inter, ntok = 6144, 12288, 16
x = mx.random.normal((ntok, h)).astype(mx.bfloat16)
def Q(o,i): 
    W=(mx.random.normal((o,i))*0.02).astype(mx.bfloat16); q=mx.quantize(W,group_size=32,bits=4,mode='mxfp4'); return q[0],q[1]
wg=Q(inter,h); wu=Q(inter,h); wd=Q(h,inter)
def mm(a,w): return mx.quantized_matmul(a,w[0],w[1],None,transpose=True,group_size=32,bits=4,mode='mxfp4')
def swiglu(dev):
    mx.set_default_device(dev)
    o = mm(nn.silu(mm(x,wg))*mm(x,wu), wd); mx.eval(o); return o.astype(mx.float32)
g,c = swiglu(mx.gpu), swiglu(mx.cpu)
print("rel-diff", (mx.mean(mx.abs(g-c))/mx.mean(mx.abs(c))).item())  # ~0.26 on Blackwell, ~0.035 on Metal
```

This points at the Blackwell CUDA mxfp4 path accumulating/rounding intermediates less precisely than Metal (or a bug specific to feeding an mxfp4-matmul result back into another mxfp4 matmul). Happy to run instrumented builds.
